### ISSUE TITLE
Timeline - Create new Tracks and Clips

### ISSUE THREAD

**Issue by lalalune** [jbilcke-hf/clapper#10]
To make Clapper a proper video creation tool, we need the ability to create content directly in the editor.

The first step in this is being able to create new clips. Here are the requirements:

1. Create new tracks
2. Set the track type from a dropdown list
3. Create clips on tracks, based on the track type
4. Drag clips on the timeline
5. Drag clips from one track to another, as long as the type is the same

Content creation within the clips will be covered in a separate PR, as this pertains to the entity editor. Primarily this feature should enable tracks and clips to be dynamically created on the timeline.

**Comment by jbilcke-hf** [comment]
I think we may need to split this one into at least two tickets 
(in that case I will provide the cash for the other ticket)

**Comment by lalalune** [comment]
I updated and simplified the scope on this.

**Comment by lalalune** [comment]
I'm doubling the bounty for this.

/boss $250

**Comment by lalalune** [comment]
I'm doubling the bounty for this.

/boss $500

**Comment by jbilcke-hf** [comment]
I've made a bult-in change to the core codebase, to have plenty of empty tracks by default

(this doesn't solve this ticket, but it will make the job easier)

I love the "Create clips on tracks, based on the track type" idea by the way!

I even wondered if this could be done in "inpaint" mode, where you could just paint into the timeline to create new cells, so the track type would act as the video instrument/synthesizer preset, and the AI would fill-in based on the context

**Comment by sendn0des** [comment]
Adding a 2nd timeline editor ticket. 
@lalalune fyi i walked devniel and j thru this with AfterEffects. I think they get it 

@devniel I added you to my ticket I made since you had mentioned wanting to tackle this.  


> I even wondered if this could be done in "inpaint" mode, where you could just paint into the timeline to create new cells, so the track type would act as the video instrument/synthesizer preset, and the AI would fill-in based on the context

@jbilcke-hf Can you describe this more? Sounds interesting ...  but not sure the use case or practically from an edit pov

**Comment by sendn0des** [comment]
We chatted more here. Possible get one extra dev hand here with Daniel.

**Comment by sendn0des** [comment]
frontend only

**Comment by pnogemane** [comment]
Hi @lalalune  is this task still available to pick up?

**Comment by kallal79** [comment]
Hi @lalalune is this task still available to pick up?

**Comment by satyamkumar420** [comment]
## 🎉 Implementation Complete!

I've implemented all the requested features for timeline track and clip creation. Here's what's been added:

### ✅ Features Implemented

1. **Create New Tracks** - `createTrack()` function with category selection
2. **Set Track Type from Dropdown** - TimelineToolbar with category selector
3. **Create Clips on Tracks** - `createClip()` function with full configuration
4. **Drag Clips on Timeline** - Enhanced existing drag functionality
5. **Drag Clips Between Tracks** - `moveSegmentToTrack()` with type validation

### 🔗 Pull Request

PR #134: https://github.com/jbilcke-hf/clapper/pull/134

### 📦 What's Included

- **3 new functions** in timeline store (createTrack, createClip, moveSegmentToTrack)
- **TimelineToolbar component** with intuitive UI for track/clip creation
- **Type validation** for drag-and-drop between tracks
- **Collision detection** to prevent overlapping clips
- **Comprehensive tests** (15+ test cases)
- **Full documentation** in TIMELINE_FEATURES.md

### 🎨 UI Features

The new toolbar provides:
- Category dropdown (VIDEO, IMAGE, DIALOGUE, MUSIC, SOUND, etc.)
- Track selector
- "New Track" button
- "New Clip" button

### 🧪 Testing

All features are fully tested and validated:
- Track creation with different categories
- Clip creation at cursor position
- Track compatibility validation
- Collision detection
- Integration scenarios

### 📚 Documentation

Complete documentation available in `TIMELINE_FEATURES.md` including:
- Usage examples
- API reference
- Technical details
- Future enhancement ideas

Ready for review! Let me know if you need any changes or have questions. 🚀

**Comment by pelegs202-design** [comment]
I reviewed the thread and saw the implementation attempt in #134.

If there is still remaining scope for #10, I can take a finalization pass focused on:
1. Identify which acceptance points are still not satisfied in current code.
2. Implement only the missing timeline behavior (track/clip creation and movement constraints).
3. Add/update regression tests for remaining gaps.
4. Open a PR with verification notes and `closes #10`.

If this direction is useful, please share the exact remaining checklist so I avoid duplicating prior work.

**Comment by yihangwu539-png** [comment]
I'll work on this bounty (Timeline - Create new Tracks and Clips). Claiming it. 👨‍💻

**Comment by yihangwu539-png** [comment]
I'll work on this bounty. Claiming it. 👨‍💻

**Comment by Tusharkhadde** [comment]
Closes #10

Implemented a scoped first pass of timeline-native authoring for this issue.

What’s now covered:
- Create new tracks directly from the timeline editor
- Set track type from a dropdown
- Create new clips on selected tracks
- Enforce type compatibility when placing clips on tracks
- Add editor toolbar controls for quick `New Track` / `New Clip` actions

Notes:
- This PR is intentionally focused on dynamic track/clip creation and timeline placement behavior.
- Clip content editing (entity/editor-side content authoring) is kept out of scope for this step, as discussed.
- Existing unrelated workspace changes were intentionally excluded to keep review focused.

**Comment by MyTH-zyxeon** [comment]
Hi maintainers - I am tracking issue #10 and noticed there are already several open implementations (#134, #136, #139, #140), with #140 looking like the freshest attempt.

To avoid adding another noisy duplicate PR, would you prefer a focused review/acceptance pass across the existing PRs first, or is there a specific missing slice you would like a narrow companion PR to cover?

I can help against this acceptance checklist:
- typed track creation from timeline controls
- clip creation on selected track at playhead/clicked time
- horizontal drag persistence
- cross-track move only when category-compatible
- overlap/collision rejection
- entity/content editor work left out of scope

**Comment by bullyopswork** [comment]
Hi @jbilcke-hf, I’m interested in helping with this bounty, but I don’t want to add another duplicate PR since there are already several open attempts.

Would a focused review/gap pass be useful here? I can compare the existing PRs against the requested timeline track/clip behavior, identify what is still missing or broken, and either post a clear acceptance checklist or open a narrow fix PR for the remaining gap.

If you prefer a specific implementation direction for tracks/clips, I can follow that instead.

**Comment by 18166714330cl-maker** [comment]
I can take a focused look at this.

Plan: inspect the current timeline model and UI, identify the smallest useful tracks/clips slice, and only open a PR if I can keep the implementation scoped with clear verification.

**Comment by MyTH-zyxeon** [comment]
Maintainer-facing acceptance synthesis for #10 / public PRs #134, #136, #139, and #140.

I am not claiming this bounty here and I am not asking maintainers to prefer a specific submitter. This is a review-assist synthesis to make the track/clip creation acceptance surface easier to compare across the open attempts.

Issue #10 asks for a frontend-only first step toward timeline-native authoring:
- create new tracks
- set the track type from a dropdown/list
- create clips on tracks based on track type
- drag clips horizontally on the timeline
- drag clips from one track to another only when the type is compatible
- keep clip content/entity editing out of scope

Visible implementation surface checked:
- #134: broad store + toolbar + docs/tests pass, but it is older and larger, with extra docs and a separate TrackControls placeholder.
- #136: compact toolbar/store implementation with build notes, but public test plan still has manual unchecked items and Copilot comments.
- #139: smaller timeline/editor pass, currently `mergeStateStatus=UNSTABLE`.
- #140: freshest timeline-native implementation, already reviewed for duplicate SOUND category, overlap policy, track type mutation on populated tracks, negative drag timing, empty-track generic clip creation, and future track-id collision risk.

Acceptance checks I would use before marking #10 complete:
1. Track creation: adding a track should update timeline state, content size/height, selected track behavior, and any persisted project data without requiring unrelated entity-editor work.
2. Track type selection: track type should be explicit, deterministic, and not silently changed after a track already contains incompatible clips.
3. Clip creation: new clips should inherit the selected track type, use a valid start/end time, preserve positive duration, and avoid generic fallback clips unless the issue owner accepts that behavior.
4. Horizontal drag: dragging left/right should preserve duration, clamp safely at zero, snap consistently if snapping is enabled, and never produce negative or end-before-start timing.
5. Cross-track drag: move should be rejected when target type is incompatible, and should also honor overlap/collision policy on the target track.
6. Collision policy: new clip creation and drag finalization should define whether overlaps are allowed. If overlaps are not allowed, both paths need the same guard and tests.
7. Track identity: new track IDs should not rely only on current array length if deletion or reordering can exist later; max-id-plus-one or another monotonic source is safer.
8. Scope control: UI should stay focused on track/clip authoring. Generating clip content, inpaint/editor behavior, or file-menu project creation should stay separate unless maintainers explicitly expand #10.
9. Verification: acceptance should include focused regression coverage for create track, choose type, create clip, drag horizontally, reject incompatible track moves, reject/allow overlap per policy, and clamp leftward drags.

Suggested merge bar:
- The selected PR is importable in the current app/timeline package layout.
- The implementation demonstrates the five issue requirements in a focused path.
- Any known edge cases above are either fixed, intentionally documented as out of scope, or accepted by maintainers.
- The branch has clear build/typecheck/test evidence without requiring a local app workflow or unrelated dependency changes.

This should help compare the existing open attempts without encouraging another duplicate PR while there are already multiple implementations in flight.

**Comment by shaiananvari8** [comment]
/boss champion 

Opened PR #145 for this issue: https://github.com/jbilcke-hf/clapper/pull/145

Scope: timeline-native track and clip authoring with typed track creation, typed clip creation, compatible cross-track moves, overlap guards, clamped horizontal drag behavior, focused helper tests, and build validation.

**Comment by ptgram4220** [comment]
Hi, I generated a read-only triage sample for this repository that may help with the duplicate bounty/PR review load around issues like #10 and #5.

It summarizes open issues that need payout/eligibility clarification, stale issues that may be ready to close or refresh, and open PRs that are missing verification notes or appear to overlap existing bounty work.

I will not post labels, reviews, closes, or repo actions without maintainer approval. If useful, I can share the markdown report here or in a gist.

**Comment by elianguitarra** [comment]
/boss champion

Opened PR #150 for this issue: https://github.com/jbilcke-hf/clapper/pull/150

Scope: focused timeline-native track and clip authoring with typed track creation, typed clip creation at the timeline cursor, target-track selection, track-category source of truth for clips created on existing tracks, timeline metrics recalculation after dynamic creation, and build/typecheck verification notes in the PR body.

**Comment by elianguitarra** [comment]
/boss champion

Opened PR #150 for this issue: https://github.com/jbilcke-hf/clapper/pull/150

Scope: focused timeline-native track and clip authoring with typed track creation, typed clip creation at the timeline cursor, target-track selection, track-category source of truth for clips created on existing tracks, timeline metrics recalculation after dynamic creation, and build/typecheck verification notes in the PR body.

**Comment by ybyc7g8d25-oss** [comment]
Hi, I opened a scoped PR for this bounty: https://github.com/jbilcke-hf/clapper/pull/152

It adds timeline controls for creating tracks and clips, wires them into the existing .clap segment flow, and adds basic drag movement with a same-type track guard.

I kept the change inside packages/timeline so it should be easier to review. Happy to adjust the UI/behavior if you want the controls exposed somewhere else in the editor.

**Comment by Errordog2** [comment]
/boss champion
Opened PR #154 for this bounty: https://github.com/jbilcke-hf/clapper/pull/154

Scope: timeline-native track and clip authoring with store APIs, direct add-track control, track category dropdown, double-click clip creation, same-category clip movement, and regression coverage.

Validation on current PR head `6ba6178`:
- `git diff --check HEAD~4..HEAD` passed
- `npx --yes bun --no-install test ./packages/timeline/test/test.tsx` passed: 7 tests, 30 assertions

**Comment by Riw3l** [comment]
Hi @jbilcke-hf and @lalalune, I'm interested in #10, but I can see many overlapping open PRs already targeting the listed requirements. Is there still a concrete missing requirement you want a new contributor to address, and is the $500 BOSS bounty still available? I'll wait for your confirmation before starting.

**Comment by elianguitarra** [comment]
Hi @jbilcke-hf and @lalalune, could you clarify the current bounty / payout status for this issue?

I had registered as `/boss champion` and opened PR #150 for this issue: https://github.com/jbilcke-hf/clapper/pull/150

Boss is showing my account as unable to enable earning bounties because my country is MX, and the account page says `Earn bounties: not yet available`. I also see the historical Boss payout/error comment on this issue mentioning @elianguitarra with an account setup / earning-bounties error.

Could you confirm whether the $500 Boss bounty is still available for the accepted/merged solution, and whether my PR #150 is eligible for any review or partial consideration? If Boss cannot pay MX accounts directly, would you be open to coordinating an alternative payout method only if my contribution is accepted as eligible?

I am not asking to bypass the maintainer decision or claim over other open submissions; I just want to avoid losing eligibility purely because Boss does not currently support payouts for my country.

**Comment by leonaaardob** [comment]
I can take a focused pass on the remaining unmet acceptance criterion for this issue, but I see several open implementation PRs already (#145, #150, #152, #154, #156, #158). Before opening another duplicate, could you confirm:

1. which specific requirement is still missing or preferred for merge;
2. whether the $500 BOSS bounty is still active and payable to a contributor in France; and
3. which existing branch/PR should be used as the integration base, if any.

I will keep any contribution scoped to the confirmed gap and disclose AI assistance.

**Comment by teddyvj** [comment]
/boss champion

Opened PR [#160](https://github.com/jbilcke-hf/clapper/pull/160) implementing the full timeline track and clip authoring workflow:

### Summary of Implementation
1. **Track Creation**: `createTrack` action generates monotonic track IDs, sets category, automatically calculates track height (preview vs. standard), and updates timeline content metrics.
2. **Track Type Selection**: Added `TimelineAuthoringToolbar` with track and clip category selectors; also added interactive category badge cycling and `+ Add Track` button directly on the timeline track header (`LeftBarTrackScale`).
3. **Clip Creation on Tracks**: `createClipOnTrack` creates clips inheriting the track category, snapping start time to grid steps, and automatically placing clips at the next available collision-free position.
4. **Horizontal Drag & Drop**: Implemented clamped horizontal movement (`snapTimeToStep`), preventing negative timestamps and maintaining duration invariants.
5. **Cross-Track Movement**: Enforces category compatibility (rejects moves between incompatible track types; unoccupied/generic tracks inherit the moved clip's category) and prevents clip overlaps on the destination track.
6. **Safety & Stability**: Added global pointer-up listener so rapid pointer movements never leave clips in a stuck dragging state.

### Verification
- `bun test packages/timeline/src/utils/timelineAuthoring.test.ts` (6/6 passing)
- `bun run --cwd packages/clap build` (clean build)
- `bun run --cwd packages/timeline build` (bundled 154 modules, 470.95 KB, declaration types cleanly emitted)
- `git diff --check` (clean, zero formatting/lint issues)
